File: exercicios/aula13/ex04_tipo_herdado/declaracoes.c
/* Exercício 04 — propagação de tipo herdado em listas de identificadores.
 *
 * Mesma gramática, mesma SDD L-atribuída e mesma saída de declaracoes.py (ver
 * a docstring de lá para as regras e a justificativa). Em C:
 *
 * - o atributo herdado L.inh é o parâmetro `inh` de parse_L;
 * - a tabela de símbolos é um vetor dinâmico em ordem de declaração, com busca
 *   linear de duplicatas (O(n) por inserção);
 * - a tabela é dona das strings de nome; os tipos são literais estáticos.
 *
 * Uso:  ./declaracoes [--trace] <arquivo | ->
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../comum/fluxo.h"

typedef struct {
    char *nome;
    const char *tipo;
    int linha, coluna;
} Simbolo;

typedef struct {
    Simbolo *v;
    size_t n, cap;
} Tabela;

static const Simbolo *tabela_procura(const Tabela *t, const char *nome) {
    for (size_t i = 0; i < t->n; i++) {
        if (strcmp(t->v[i].nome, nome) == 0) return &t->v[i];
    }
    return NULL;
}

static void tabela_insere(Tabela *t, const Token *tok, const char *tipo) {
    if (t->n == t->cap) {
        t->cap = t->cap ? t->cap * 2 : 8;
        t->v = realoca(t->v, t->cap * sizeof(Simbolo));
    }
    t->v[t->n++] = (Simbolo){copia(tok->lexeme), tipo, tok->line, tok->column};
}

static void tabela_free(Tabela *t) {
    for (size_t i = 0; i < t->n; i++) free(t->v[i].nome);
    free(t->v);
}

typedef struct {
    Fluxo *fl;
    int trace;
    Tabela tabela;
    char **erros;
    size_t n_erros;
} Declaracoes;

static void regra(const Declaracoes *d, const char *prod, const char *acao) {
    if (d->trace) printf("%-14s%s\n", prod, acao);
}

/* Ação semântica addtype(id, L.inh). */
static void addtype(Declaracoes *d, const Token *tok, const char *tipo, const char *prod) {
    const Simbolo *anterior = tabela_procura(&d->tabela, tok->lexeme);
    if (anterior) {
        char *acao = formata("addtype(%s, L.inh) -> redeclaração", tok->lexeme);
        regra(d, prod, acao);
        free(acao);
        d->erros = realoca(d->erros, (d->n_erros + 1) * sizeof(char *));
        d->erros[d->n_erros++] = formata(
            "Erro semântico na linha %d, coluna %d: redeclaração de '%s' (declarado na linha %d, coluna %d).",
            tok->line, tok->column, tok->lexeme, anterior->linha, anterior->coluna);
        return;
    }
    char *acao = formata("addtype(%s, L.inh) -> %s : %s", tok->lexeme, tok->lexeme, tipo);
    regra(d, prod, acao);
    free(acao);
    tabela_insere(&d->tabela, tok, tipo);
}

/* Consome um token do tipo pedido; senão imprime o erro e devolve NULL. */
static const Token *exige(Fluxo *fl, TokenType tipo, const char *esperado) {
    if (fluxo_verifica(fl, tipo)) return fluxo_proximo(fl);
    erro_sintaxe(fluxo_atual(fl), esperado);
    return NULL;
}

/* T -> int | float | char; devolve T.type (NULL em erro de sintaxe). */
static const char *parse_T(Declaracoes *d) {
    const Token *t = fluxo_atual(d->fl);
    const char *tipo = t->type == TOK_KW_INT     ? "int"
                       : t->type == TOK_KW_FLOAT ? "float"
                       : t->type == TOK_KW_CHAR  ? "char"
                                                 : NULL;
    if (!tipo) {
        erro_sintaxe(t, "tipo (int, float ou char)");
        return NULL;
    }
    fluxo_proximo(d->fl);
    char *prod = formata("T -> %s", tipo), *acao = formata("T.type = %s", tipo);
    regra(d, prod, acao);
    free(prod), free(acao);
    return tipo;
}

/* L -> id { , id }; `inh` é o atributo herdado L.inh. */
static int parse_L(Declaracoes *d, const char *inh) {
    const Token *id = exige(d->fl, TOK_IDENT, "identificador");
    if (!id) return ERRO_SINTAXE;
    addtype(d, id, inh, "L -> id");
    while (fluxo_aceita(d->fl, TOK_COMMA)) {
        if (!(id = exige(d->fl, TOK_IDENT, "identificador"))) return ERRO_SINTAXE;
        addtype(d, id, inh, "L -> L , id");
    }
    return OK;
}

static int parse_D(Declaracoes *d) {
    const char *tipo = parse_T(d);
    if (!tipo) return ERRO_SINTAXE;
    char *acao = formata("L.inh = T.type = %s", tipo);
    regra(d, "D -> T L ;", acao);
    free(acao);
    if (parse_L(d, tipo) != OK) return ERRO_SINTAXE;
    return exige(d->fl, TOK_SEMICOLON, "',' ou ';'") ? OK : ERRO_SINTAXE;
}

int main(int argc, char **argv) {
    int trace = 0, n_args = 0;
    const char *arquivo = NULL;
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--trace") == 0) trace = 1;
        else arquivo = argv[i], n_args++;
    }
    if (n_args != 1) {
        fprintf(stderr, "uso: %s [--trace] <arquivo | ->\n", argv[0]);
        return ERRO_USO;
    }
    Fluxo fl;
    int codigo = fluxo_abrir(&fl, arquivo);
    if (codigo != OK) return codigo;

    Declaracoes d = {&fl, trace, {0}, NULL, 0};
    while (codigo == OK && !fluxo_verifica(&fl, TOK_EOF)) {
        codigo = parse_D(&d);
    }
    if (codigo == OK) {
        printf("tabela de símbolos:\n");
        for (size_t i = 0; i < d.tabela.n; i++) {
            const Simbolo *s = &d.tabela.v[i];
            printf("  %s : %s (linha %d, coluna %d)\n", s->nome, s->tipo, s->linha, s->coluna);
        }
        for (size_t i = 0; i < d.n_erros; i++) fprintf(stderr, "%s\n", d.erros[i]);
        codigo = d.n_erros ? ERRO_SEMANTICO : OK;
    }
    for (size_t i = 0; i < d.n_erros; i++) free(d.erros[i]);
    free(d.erros);
    tabela_free(&d.tabela);
    fluxo_fechar(&fl);
    return codigo;
}

File: exercicios/aula13/ex04_tipo_herdado/declaracoes.py
"""Exercício 04 — propagação de tipo herdado em listas de identificadores.

Gramática e SDD L-atribuída:

    P -> D*
    D -> T L ;        L.inh = T.type
    T -> int          T.type = int        (idem float, char)
    L -> L1 , id      L1.inh = L.inh;  addtype(id, L.inh)
    L -> id           addtype(id, L.inh)

`T.type` é sintetizado; `L.inh` é herdado — vem do irmão à esquerda (T) e
desce pela lista. A SDD é L-atribuída porque todo atributo herdado depende só
de atributos do pai ou de irmãos **à esquerda**: L.inh usa T.type (à esquerda
de L em D -> T L ;) e L1.inh usa L.inh (do pai). Nada depende de algo à
direita, então uma única passada da esquerda para a direita — a ordem em que o
parser descendente lê os tokens — calcula tudo.

Em código, "herdado" vira **parâmetro**: `parse_L(tipo)` recebe o valor de
L.inh de quem a chama. A recursão à esquerda de L vira um laço
(`L -> id { , id }`), com o mesmo tipo repassado a cada identificador.

A tabela de símbolos guarda nome, tipo, linha e coluna, e lista em ordem de
declaração. Redeclarar um nome é erro semântico: a análise continua (para
mostrar todos os erros), a primeira declaração fica valendo e o código de
saída é 4.

Uso:  python declaracoes.py [--trace] <arquivo | ->
"""

import sys
from dataclasses import dataclass
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "comum"))
import fluxo as F  # noqa: E402

TIPOS = {"KW_INT": "int", "KW_FLOAT": "float", "KW_CHAR": "char"}


@dataclass
class Simbolo:
    nome: str
    tipo: str
    linha: int
    coluna: int


class TabelaSimbolos:
    def __init__(self):
        self.simbolos: list[Simbolo] = []  # ordem de declaração
        self._por_nome: dict[str, Simbolo] = {}

    def procura(self, nome: str):
        return self._por_nome.get(nome)

    def insere(self, simbolo: Simbolo):
        self.simbolos.append(simbolo)
        self._por_nome[simbolo.nome] = simbolo


class Declaracoes:
    def __init__(self, fl: F.Fluxo, trace: bool):
        self.fl = fl
        self.trace = trace
        self.tabela = TabelaSimbolos()
        self.erros: list[str] = []

    def _regra(self, producao: str, acao: str):
        if self.trace:
            print(f"{producao:<14}{acao}")

    def parse_P(self):
        while not self.fl.verifica("EOF"):
            self.parse_D()

    def parse_D(self):
        tipo = self.parse_T()
        self._regra("D -> T L ;", f"L.inh = T.type = {tipo}")
        self.parse_L(tipo)
        self.fl.exige("SEMICOLON", "',' ou ';'")

    def parse_T(self) -> str:
        tok = self.fl.atual()
        if tok.type not in TIPOS:
            raise F.ErroSintaxe(tok, "tipo (int, float ou char)")
        self.fl.proximo()
        tipo = TIPOS[tok.type]
        self._regra(f"T -> {tipo}", f"T.type = {tipo}")
        return tipo

    def parse_L(self, inh: str):
        """L -> id { , id }; `inh` é o atributo herdado L.inh."""
        self.addtype(self.fl.exige("IDENT", "identificador"), inh, "L -> id")
        while self.fl.aceita("COMMA"):
            self.addtype(self.fl.exige("IDENT", "identificador"), inh, "L -> L , id")

    def addtype(self, tok, tipo: str, producao: str):
        """Ação semântica: registra id com o tipo herdado."""
        anterior = self.tabela.procura(tok.lexeme)
        if anterior is not None:
            self._regra(producao, f"addtype({tok.lexeme}, L.inh) -> redeclaração")
            self.erros.append(
                F.erro_semantico(
                    tok.line,
                    tok.column,
                    f"redeclaração de '{tok.lexeme}' (declarado na linha {anterior.linha}, "
                    f"coluna {anterior.coluna})",
                )
            )
            return
        self._regra(producao, f"addtype({tok.lexeme}, L.inh) -> {tok.lexeme} : {tipo}")
        self.tabela.insere(Simbolo(tok.lexeme, tipo, tok.line, tok.column))


def main(argv: list[str]) -> int:
    F.configurar_saida()
    trace = "--trace" in argv[1:]
    args = [a for a in argv[1:] if a != "--trace"]
    if len(args) != 1:
        print("uso: python declaracoes.py [--trace] <arquivo | ->", file=sys.stderr)
        return F.ERRO_USO

    fl, codigo = F.abrir(args[0])
    if fl is None:
        return codigo
    d = Declaracoes(fl, trace)
    try:
        d.parse_P()
    except F.ErroSintaxe as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SINTAXE

    print("tabela de símbolos:")
    for s in d.tabela.simbolos:
        print(f"  {s.nome} : {s.tipo} (linha {s.linha}, coluna {s.coluna})")
    for e in d.erros:
        print(e, file=sys.stderr)
    return F.ERRO_SEMANTICO if d.erros else F.OK


if __name__ == "__main__":
    sys.exit(main(sys.argv))
